### PR TITLE
Exp/serializer refactoring

### PR DESCRIPTION
**Pull request by fulmicoton**

### CODE REVIEW

File: src/common/composite_file.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
+use std::io::Write;
+use common::CountingWriter;
+use std::collections::HashMap;
+use schema::Field;
+use common::VInt;
+use std::io;
+use directory::ReadOnlySource;
+use common::BinarySerializable;
+
+pub struct CompositeWrite<W: Write> {
```
Comment (fulmicoton) [inline src/common/composite_file.rs]: doc required

File: src/common/composite_file.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    pub fn for_field(&mut self, field: Field) -> &mut CountingWriter<W> {
+        let offset = self.write.written_bytes();
+        assert!(!self.offsets.contains_key(&field));
+        self.offsets.insert(field, offset);
+        &mut self.write
+    }
+
+    pub fn close(&mut self) -> io::Result<()> {
```
Comment (fulmicoton) [inline src/common/composite_file.rs]: `close(self)`

File: src/compression/stream.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
+use compression::BlockDecoder;
+use compression::NUM_DOCS_PER_BLOCK;
+use compression::compressed_block_size;
+use directory::SourceRead;
+
+pub struct CompressedIntStream {
```
Comment (fulmicoton) [inline src/compression/stream.rs:6]: Add doc

File: src/core/searcher.rs [hunk 4]
```diff
@@ -46,7 +47,12 @@ impl Searcher {
     pub fn doc_freq(&self, term: &Term) -> u32 {
         self.segment_readers
             .iter()
-            .map(|segment_reader| segment_reader.doc_freq(term))
+            .map(|segment_reader| {
+                segment_reader
+                    .field_reader(term.field())
+                    .unwrap() // TODO error handling
```
Comment (fulmicoton) [inline src/core/searcher.rs:52]: Error handling

File: src/core/searcher.rs [hunk 5]
```diff
@@ -65,20 +71,46 @@ impl Searcher {
         query.search(self, collector)
     }
 
+    pub fn field(&self, field: Field) -> Result<FieldSearcher> {
```
Comment (fulmicoton) [inline src/core/searcher.rs]: See if we can remove this or at least make FieldSearcher private

**Review by KodrAus — COMMENTED** [review]
Just left a few random comments and nits.

**Pull request merged**